src/hardware/siglent-sds/protocol.c

```diff
@@ -364,76 +411,127 @@ static int siglent_sds_get_digital(const struct sr_dev_inst *sdi, struct sr_chan
 	for (l = sdi->channels; l; l = l->next) {
 		ch = l->data;
 		samples_index = 0;
-		if (ch->type != SR_CHANNEL_LOGIC)
-			continue;
-		if (!ch->enabled)
-			continue;
-		if (sr_scpi_send(sdi->conn, "D%d:WF? DAT2", ch->index) != SR_OK)
-			return SR_ERR;
-		if (sr_scpi_read_begin(scpi) != SR_OK)
-			return TRUE;
-		len = sr_scpi_read_data(scpi, buf, -1);
-		if (len < 0)
-			return TRUE;
-		len -= 15;
-		buffdata = g_array_sized_new(FALSE, FALSE, sizeof(uint8_t), len);
-		buf += 15; /* Skipping the data header. */
-		g_array_append_vals(buffdata, buf, len);
-		tmp_samplebuf = g_array_sized_new(FALSE, FALSE, sizeof(uint8_t), len); /* New temp buffer. */
-		for (uint64_t cur_sample_index = 0; cur_sample_index < devc->memory_depth_digital; cur_sample_index++) {
-			char sample = (char)g_array_index(buffdata, uint8_t, cur_sample_index);
-			for (int ii = 0; ii < 8; ii++, sample >>= 1) {
-				if (ch->index < 8) {
-					channel_index = ch->index;
-					if (data_low_channels->len <= samples_index) {
-						tmp_value = 0; /* New sample. */
-						low_channels = TRUE; /* We have at least one enabled low channel. */
-					} else {
-						/* Get previous stored sample from low channel buffer. */
-						tmp_value = g_array_index(data_low_channels, uint8_t, samples_index);
+		if (ch->type == SR_CHANNEL_LOGIC) {
```

Comment:
this section is complex and hard to review (I am skipping over it at this time)
Also if there's that many different sub-cases with "if model is SDS2000", it starts making sense to have a separate function entirely.